### PR TITLE
Add raster visualization with envelope overlay for taste delivery

### LINKED ISSUE #27: Add visualization (raster of predictions around taste delivery) and overlay of envelope with predictions
#%% IMPORTS
import pandas as pd
import os
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.colors import ListedColormap
import matplotlib as mpl
import seaborn as sns
from sklearn.preprocessing import StandardScaler
from umap import UMAP
from sklearn.decomposition import PCA
from tqdm import tqdm

#%% DEFINE FUNCTIONS
def return_pred_array(taste_frame):
    """
    Given a taste_frame, return a 2D array of predictions
    
    Inputs:
        taste_frame : pd.DataFrame

    Outputs:
        pred_array : np.array
            2D array with shape (n_trials, max_time)
    """

    assert len(taste_frame.taste.unique()) == 1
    assert len(taste_frame.basename.unique()) == 1
    n_trials = taste_frame.trial.max() + 1
    max_time = np.max([x for y in taste_frame.segment_bounds for x in y] )
    # Round up to nearest 1000
    max_time = int(np.ceil(max_time/100) * 100)
    pred_array = np.zeros((n_trials, max_time))
    pred_array[:] = np.nan 
    for _, this_row in taste_frame.iterrows():
        this_trial = this_row.trial
        this_bounds = this_row.segment_bounds
        this_pred = this_row.xgb_pred_code
        pred_array[this_trial, this_bounds[0]:this_bounds[1]] = this_pred
    return pred_array

#%% LOAD DATA & PLOT
artifact_dir = '/media/cmazzio/large_data/CM77/CM77_CTATest2_h2o_nacl_lowqhcl_highqhcl_250627_103549/emg_classifier' 

xgb_pred_plot_dir = os.path.join(artifact_dir, 'pipeline_test_plots')
if not os.path.exists(xgb_pred_plot_dir):
    os.makedirs(xgb_pred_plot_dir)

bsa_event_map = {
        0 : 'no movement',
        1 : 'gape',
        2 : 'MTMs',
        }
event_color_map = {
        0 : '#D1D1D1',
        1 : '#EF8636',
        2 : '#3B75AF',
        }
inv_bsa_event_map = {v: k for k, v in bsa_event_map.items()}

predicted_df = pd.read_pickle('/media/cmazzio/large_data/CM77/CM77_CTATest2_h2o_nacl_lowqhcl_highqhcl_250627_103549/emg_classifier/CM77_CTATest2_emg_classifier_segments.pkl')

predicted_df.reset_index(inplace=True)
predicted_df['xgb_pred_code'] = predicted_df['pred_names'].map(inv_bsa_event_map)
# feature_names = open(os.path.join(artifact_dir, 'all_datasets_feature_names.txt')).read().split('\n') 
# =============================================================================
# predicted_df['animal_num'] = predicted_df.basename.str.split('_').str[0]
# predicted_df['animal_code'] = predicted_df.animal_num.astype('category').cat.codes
# predicted_df['session_code'] = predicted_df.basename.astype('category').cat.codes
# =============================================================================

cmap = ListedColormap(list(event_color_map.values()), name = 'NBT_cmap')

############################################################
# 1- Features UMAP colored by predicted class
# Create UMAP
predicted_df = predicted_df.sort_values(
    ['xgb_pred_code', 
    #'animal_code', 
    # 'session_code'
     ]
         )
pred_dict = {}
# for session_name, session_frame in tqdm(list(predicted_df.groupby('basename'))):
for taste_name, taste_frame in list(predicted_df.groupby('taste')):
    taste_frame['basename'] = 'test'
    pred_array = return_pred_array(taste_frame)
    this_pred_dict = dict(
            pred_array = pred_array,
            taste_name = taste_name,
            # session_name = session_name,
            )
    pred_dict[f'{taste_name}'] = this_pred_dict

pred_frame = pd.DataFrame(pred_dict).T

# Plot
# xgb_pred_plot_dir = os.path.join(plot_dir, 'pipeline_test_plots', 'xgb')

# Create segmented colormap

# Plot one session at a time
# for i, this_xgb_pred  in enumerate(xgb_pred_array_list):
    
pred_frame['session_name'] = 'test'
for session_name, session_frame in list(pred_frame.groupby('session_name')):
    this_xgb_pred = session_frame.pred_array.to_numpy()
    this_tastes = session_frame.taste_name.tolist()

    fig, ax = plt.subplots(len(this_tastes),1,sharex=True,sharey=True,
                           figsize=(5,10))
    for taste in range(4):
        try:
            this_array = this_xgb_pred[taste]
            max_trials = this_array.shape[0]
            x_vec = np.arange(this_array.shape[1])
            im = ax[taste].pcolormesh(
                    x_vec, np.arange(max_trials), 
                    this_xgb_pred[taste],
                      cmap=cmap,vmin=0,vmax=2,)
            ax[taste].set_ylabel(f'{this_tastes[taste]}' + '\nTrial #')
        except:
            print(f'{session_name} {taste} ind not found')
    ax[0].set_title('XGB')
    ax[-1].set_xlabel('Time (ms)')
    cbar_ax = fig.add_axes([0.98, 0.15, 0.02, 0.7])
    cbar = fig.colorbar(im, cax=cbar_ax)
    cbar.set_ticks([0.5,1,1.5])
    cbar.set_ticklabels(['nothing','gape','MTMs'])
    plt.tight_layout()
    plt.subplots_adjust(top=0.9)
    fig.suptitle(session_name)
    fig.savefig(os.path.join(xgb_pred_plot_dir, session_name + '_xgb_bsa_pred_test.png'),
                bbox_inches='tight', dpi = 300)
    plt.close(fig)

### PR DESCRIPTION
**Pull request by abuzarmahmood**
## Summary
This PR addresses issue #27 by adding a new visualization function that creates raster plots of predictions around taste delivery with EMG envelope overlay.

### New Function: `generate_raster_with_envelope()`
Added to `src/visualize.py` - creates a comprehensive visualization showing:
- **Left panel**: EMG envelope signals for each trial with colored prediction overlays
- **Right panel**: Raster plot showing movement classifications over time
- Each taste gets its own row of subplots
- Trials are displayed with vertical offsets for clarity
- Color-coded movement types (nothing, gape, MTMs)

### Key Features
- Normalizes EMG signals using median absolute deviation for robust visualization
- Overlays predictions as colored regions on the envelope traces
- Provides both detailed envelope view and compact raster view side-by-side
- Supports optional taste names and session naming
- Handles single or multiple tastes automatically

### Integration
- Updated `src/run_flow.py` to import and use the new function
- Added example usage that generates the visualization automatically

## Testing
- Code follows existing patterns in the codebase
- Uses the same color scheme as other visualization functions
- Compatible with existing data structures (segments_frame, raw_emg)

## Resolves
Closes #27

### CODE REVIEW

**Comment by abuzarmahmood** [comment]
## Implementation Details

### Visualization Approach
The `generate_raster_with_envelope()` function creates a dual-panel visualization for each taste:

1. **Envelope Panel (Left, 3x width)**:
   - Displays raw EMG envelope traces for all trials
   - Each trial is vertically offset by 10 units for visual separation
   - EMG signals are normalized using median absolute deviation (MAD) for robustness against outliers
   - Predictions are overlaid as semi-transparent colored regions behind the envelope traces
   - Color coding: Gray (nothing), Orange (gape), Blue (MTMs)

2. **Raster Panel (Right, 1x width)**:
   - Compact raster plot showing predictions as colored pixels
   - Each row represents a trial, columns represent time
   - Uses pcolormesh for efficient rendering of large datasets
   - Same color scheme as envelope panel for consistency

### Technical Considerations
- **Normalization**: Uses MAD instead of standard deviation for better handling of EMG artifacts
- **Scalability**: Handles variable numbers of tastes and trials automatically
- **Layout**: Uses gridspec for flexible subplot sizing (3:1 ratio)
- **Color consistency**: Reuses the existing color map from other visualization functions
- **Edge cases**: Handles single-taste datasets by reshaping axes array

### Usage Example
```python
from visualize import generate_raster_with_envelope

fig, axes = generate_raster_with_envelope(
    segments_frame=segment_frame,
    raw_emg=env,
    session_name="My Session",
    taste_names=["Water", "NaCl", "Citric Acid", "Quinine"]
)
plt.show()
```

**Comment by abuzarmahmood** [comment]
## Comparison with Existing Visualizations

This new function complements the existing visualization tools:

### `generate_raster_plot()`
- **Purpose**: Shows only the raster of predictions
- **Use case**: Quick overview of classification results
- **New function adds**: EMG envelope context alongside the raster

### `plot_env_pred_overlay()`
- **Purpose**: Grid of all trials showing EMG with prediction overlays
- **Use case**: Detailed inspection of individual trials
- **New function adds**: Compact raster view + better layout for comparing across tastes

### `generate_raster_with_envelope()` (NEW)
- **Purpose**: Combined view of envelope traces and raster plot
- **Use case**: Understanding predictions in context of EMG signal patterns
- **Advantages**: 
  - Side-by-side comparison of detailed and compact views
  - Better for presentations and publications
  - Easier to spot patterns across trials and tastes
  - More efficient use of figure space

The new visualization is particularly useful for:
- Validating that predictions align with visible EMG patterns
- Identifying systematic differences between tastes
- Presenting results in papers/talks where space is limited
- Quick quality checks during analysis

**Comment by abuzarmahmood** [comment]
Already implemented. Closing.
